### PR TITLE
Add Image Access Chapter

### PR DESCRIPTION
**Pull request by spencer-lunarg**
closes https://github.com/KhronosGroup/SPIRV-Guide/issues/24

### CODE REVIEW

**Review by alan-baker — COMMENTED** [review]
https://github.com/KhronosGroup/SPIRV-Headers/issues/448 raising some common issues about images. I think it might make sense to try and explain images as handles in this chapter. Answering the question of what it means to load an image.

File: chapters/image_accceses.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# Image Accesses
+
+SPIR-V has various [Image Instructions](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_image_instructions) that will interact with `OpTypeImage`. This chapter aims to give some extra context around these instructions.
```
Comment (mmoult) [inline chapters/image_accceses.md:3]: I would recommend removing the word "will" ->

"SPIR-V has various Image Instructions that ~~will~~ interact with `OpTypeImage`."

I think it is accurate to say the images work now, so the present tense is confusing.

File: chapters/image_accceses.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+> There is an [example.frag](./examples/image_accesses/example.frag) (and [example.spv](./examples/image_accesses/example.spv)) GLSL shader that has an example of various ways images can be accessed in SPIR-V.
+
+# Images as Handles
+
+An `OpTypeImage` can be thought of as a handle the underlying data. To help illustrate this, here is a small program that writes to an image
```
Comment (mmoult) [inline chapters/image_accceses.md:9]: Looks like you are missing a preposition in this sentence.

"An `OpTypeImage` can be thought of as a handle *to*/*for*/*of* the underlying data."

File: chapters/image_accceses.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+```glsl
+layout(set = 0, binding = 1, rgba8) uniform writeonly image2D image;
+void main() {
+    imageStore(image, ivec2(0, 0), vec4(1.0));
+}
+```
+
+Taking a closer look at the `OpVariable` we see it is of type [UniformConstant](../chapters/storage_class.md#uniformconstant) which is suppose to be "read-only" (yet there is a clear `OpImageWrite` occuring here).
+
+This is possible because the handle to the image is read-only, but the underlying texel data of the image is mutable.
```
Comment (mmoult) [inline chapters/image_accceses.md:20]: I would appreciate clarification on this point. What would it mean for an image variable to have a mutable storage class (such as `Uniform`)? Is that legal?

I still feel some confusion about which modifiers apply to which data.

- `NonReadable` and `NonWritable` apply to the underlying texel data.
- Storage class applies to the metadata, but I got the idea from Alan that the fields in `OpTypeImage` are all immutable, regardless of storage class. If that is true, what does the storage class really say about the image variable?

File: chapters/image_accceses.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  // other metadata ...
+  void* texels; // data of the image (which is mutable)
+}
+```
+
+The above `OpVariable` is read-only because we are not allowed to adjust its format, width, or multisampling state ever.
+
+# Sampled Images
+
+When using a sampler, you must have a `OpTypeSampledImage` object, this will be directly pointing to a `OpTypeImage` object.
```
Comment (mmoult) [inline chapters/image_accceses.md:41]: Consider using "an" when talking about a single `Op...`. e.g.:

"When using a sampler, you must have *an* `OpTypeSampledImage` object..."

File: chapters/image_accceses.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+While the `OpImage*Sample*` takes float coordinates and does a lookup to get the texel, the `OpImageFetch` takes direct image coordinates to the exact texel to return.
+
+This means an `OpImageFetch` will directly access through an `OpImage` and doesn't need an `OpSampledImage` object.
+
+# Non-sampled image
+
+When not using a sampler (ex. storage image), you can directly use an `OpTypeImage` object.
+
+> Note you will **not** be using an `OpImage` when accessing an image without a sampler. The `OpImage` is designed to point to the image of the sampler/image combination.
```
Comment (mmoult) [inline chapters/image_accceses.md:59]: This seems important, but I don't think I understand the point you are making. Please articulate the difference between an `OpTypeImage` object and an `OpImage`.

The way I see it currently (so you can correct any misconceptions I may have) is that the `OpTypeImage` instruction defines an image type. An object defined/created with that image type is an image / `OpImage`.

I am looking at a couple examples I have, and `OpImageRead` and `OpImageWrite` look to me like they use a loaded image, not an image type.

File: chapters/image_accceses.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+The `OpImageRead` and `OpImageWrite` will directly go to the `OpLoad` into the variable.
+
+![image_access_storage_image.png](../images/image_access_storage_image.png)
+
+> The `OpImageWrite` calls `OpLoad` but this is **not** an image "read access". The `OpLoad` just grabs the reference, but there is only a "write access" here on the image.
```
Comment (mmoult) [inline chapters/image_accceses.md:65]: If I may suggest alternate wording:

"The `OpImageWrite` and `OpImageRead` instructions require an image (loaded by a previous `OpLoad`) as their third and second operand, respectively. However, it is important to note that loading an image does **not** load the texels themselves. Rather, it loads the image handle, among which is a pointer to the relevant texels. As such, loading the image is **not** forbidden for an image marked as `NonReadable`."

Obviously you don't need to copy what I put, but I think the current wording is confusing because I don't understand how an `OpImageWrite` calls an `OpLoad` and the description of "grab[bing] the reference" is not very clear.

File: chapters/image_accceses.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+# Image Queries
+
+There are some `OpImageQuery*` instructions that are designed to directly access an `OpTypeImage` object. The image itself is not accessed.
+
+## OpImageTexelPointer
+
+Before you do an operation such as an atomic image load, you need to get a pointer for the memory.
+
+The `OpImageTexelPointer` allows you take a texel inside an image, and create a pointer to it which can be accessed by an atomic image operation.
```
Comment (mmoult) [inline chapters/image_accceses.md:75]: I think you mistakenly forgot the word "to":

"The `OpImageTexelPointer` allows you *to* take a texel ..."

**Review by mmoult — COMMENTED** [review]
I recognize that this feedback is entirely unsolicited, but I thought you SPIR-V experts might appreciate the perspective of someone who is still learning this. Cheers!

**Comment by spencer-lunarg** [comment]
> I recognize that this feedback is entirely unsolicited, but I thought you SPIR-V experts might appreciate the perspective of someone who is still learning this

wrong, the feedback is VERY well appreciated and this guide is quite literally designed for someone learning SPIR-V so seem like the perfect audience to me!

File: chapters/image_accceses.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+To access the sampled image you will use a `OpImage*Sample*` instruction (ex. `OpImageSampleImplicitLod`) to get a `OpSampledImage`. (It can also use a `OpImage*Gather` as well)
+
+![image_access_sampled_image.png](../images/image_access_sampled_image.png)
+
+## Fetch
+
+There are also `OpImageFetch` (and `OpImageSparseFetch`) instructions which work similar as `OpImage*Sample*`.
+
+While the `OpImage*Sample*` takes float coordinates and does a lookup to get the texel, the `OpImageFetch` takes direct image coordinates to the exact texel to return.
```
Comment (bashbaug) [inline chapters/image_accceses.md:51]: Suggestions and corrections:

* The `OpImage*Sample*` instructions can take integer coordinates in addition to float coordinates.
* What does "direct image coordinates" mean?  I _think_ this is describing the difference between normalized and un-normalized coordinates.  If that's the case, maybe we should use those terms, since they're used in the SPIR-V spec?

File: images/image_access_storage_image.png
Comment (bashbaug) [inline images/image_access_storage_image.png:1]: This diagram would be clearer if it used `OpImageRead` and `OpImageWrite` instead of `OpRead` and `OpWrite`.

File: chapters/image_accceses.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+```swift
+%1 = OpImageTexelPointer %OpTypePointer %atomic_image %coor %sample
+     OpAtomicStore %1 %_ %_ %scalar_value
+
+%2 = OpLoad %OpTypeImage %image
+     OpImageWrite %2 %coor %texel
+```
+
+While the `OpLoad` returns a `OpTypeImage` that we store the new texel, with `OpImageTexelPointer` we get a direct pointer to the texel data and can store scalar value like `OpConstant %int 0`. This is posssible because the `Coordinate` operand is supplied in the `OpImageTexelPointer` instead aftewards in the actual "write" operation.
```
Comment (bashbaug) [inline chapters/image_accceses.md:89]:
```suggestion
While the `OpLoad` returns a `OpTypeImage` that we store the new texel, with `OpImageTexelPointer` we get a direct pointer to the texel data and can store scalar value like `OpConstant %int 0`. This is posssible because the `Coordinate` operand is supplied in the `OpImageTexelPointer` instead afterwards in the actual "write" operation.
```

Comment (bashbaug) [reply chapters/image_accceses.md:89]: Might be worth saying explicitly (I think this is true?): you're not expected to do any pointer math on the pointer returned by `OpImageTexelPointer`.  To get a pointer to a different texel, you should call `OpImageTexelPointer` again, with a different coordinate.

**Review by bashbaug — COMMENTED** [review]
General comment: I'm not sure how best to convey this, but I usually have trouble remembering what the difference is between the different SPIR-V image read and write instructions, or from the other direction, which SPIR-V image read or write instruction is best for a specific use.  Two possible ideas:

1. Have a decision tree / flow chart with questions about image usage, that eventually uniquely leads to each SPIR-V instruction.  For example, "Are you reading an image or writing an image?"  "Do you have a sampler?"  etc.
2. Have one or several tables where each row is a SPIR-V image access instruction, and each column describes some simple properties of the instruction, and each table cell is either Y (the image access instruction has that property), N (the image access instruction does not have that property), or a short 2-3 word description.

What do you think?

File: chapters/image_accceses.md (hunk 5 above)
Comment (bashbaug) [reply chapters/image_accceses.md:59]: This is a great comment.  I think `OpImage` imprecisely used in this sentence.

Note that `OpImage` is a SPIR-V instruction that extracts an `OpTypeImage` from an `OpTypeSampledImage`.  So, a good mental model is that `OpImage` is describing an action, rather than the input to or the output from an action.

File: chapters/image_accceses.md (hunk 8 above)
Comment (spencer-lunarg) [reply chapters/image_accceses.md:51]: > describing the difference between normalized and un-normalized coordinates.

yes, you are right, let me try rewording this

File: chapters/image_accceses.md (hunk 5 above)
Comment (spencer-lunarg) [reply chapters/image_accceses.md]: Thanks @bashbaug for the better way to describe this, updated with better wording

File: chapters/image_accceses.md (hunk 6 above)
Comment (spencer-lunarg) [reply chapters/image_accceses.md]: removed "reference" and used "handle" to be consistent.

added new diagram to hopefully also help explain what is going on

File: chapters/image_accceses.md (hunk 9 above)
Comment (spencer-lunarg) [reply chapters/image_accceses.md]: aww, never thought about that, added!

Comment (mmoult) [inline chapters/image_accceses.md]: See my comment above. I don't believe that `OpLoad` returns an `OpTypeImage`. Instead, it returns an Image object (object of type `OpTypeImage`).

**Review by mmoult — CHANGES_REQUESTED** [review]

File: chapters/image_accceses.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+The `OpImageRead` and `OpImageWrite` will directly go to the `OpLoad` into the variable.
+
+![image_access_storage_image.png](../images/image_access_storage_image.png)
+
+> The `OpImageWrite` calls `OpLoad` but this is **not** an image "read access". The `OpLoad` just grabs the reference, but there is only a "write access" here on the image.
+
+# Image Queries
+
+There are some `OpImageQuery*` instructions that are designed to directly access an `OpTypeImage` object. The image itself is not accessed.
```
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply chapters/image_accceses.md:104]: the "directly" here was ment to say "you load the variable" ... tried to reword it to be more consistent

File: chapters/image_accceses.md (hunk 9 above)
Comment (spencer-lunarg) [reply chapters/image_accceses.md]: agreed, used "image object of `OpTypeImage`" everywhere now

File: chapters/image_accceses.md [hunk 11]
```diff
@@ -0,0 +1,146 @@
+# Image Accesses
+
+SPIR-V has various [Image Instructions](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_image_instructions) that interact with `OpTypeImage`. This chapter aims to give some extra context around these instructions.
+
+> There is an [example.frag](./examples/image_accesses/example.frag) (and [example.spv](./examples/image_accesses/example.spv)) GLSL shader that has an example of various ways images can be accessed in SPIR-V.
+
+# Coordinate Systems
+
+It is important to know there are 3 texel coordinate systems used for images:
```
Comment (mmoult) [inline chapters/image_accceses.md:9]: I am not sure where it fits (and perhaps it doesn't fit in this document), but just as the coordinates can use ints and floats (which may be normalized or not), I was curious about the texel data (which I know can be ints or floats). Is there any way to know whether float texel data is normalized?

Comment (spencer-lunarg) [reply chapters/image_accceses.md:9]: I assume the Format itself could define that, like if you look at `VK_FORMAT_A2B10G10R10_UNORM_PACK32` it is defined as

>  32-bit packed unsigned normalized format

and from there is mapped to the SPIR-V format 

https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#spirvenv-image-formats

image

File: chapters/image_accceses.md [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+- Normalized (floats)
+- Unnormalized (floats)
```
Comment (bashbaug) [inline chapters/image_accceses.md:12]: Possible upgrade for a future update: consider briefly describing the difference between "normalized" and "unnormalized" coordinates, and a pointer where to learn more.

File: chapters/image_accceses.md [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+- Integers
+
+The SPIR-V spec documents which instructions are allowed to use which coordinate system.
+The [Vulkan Spec Texel Coordinate Systems Chapter](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#textures-texel-coordinate-systems) is a good resource to learn more about this.
+
+When using a `vec4` (4 wide vector) as a coordinate, the API spec defines which each vector element points to. As an example, in Vulkan, when using Unnormalized coordinates, the `vec4.z` element is the value of the `array layer`.
```
Comment (bashbaug) [inline chapters/image_accceses.md:18]: Possible future upgrades here:

* I think there is a bug in "defines which each vector element", maybe "which" should be "what"?
* "points to" is potentially confusing, since the coordinates are not pointers.  Maybe "represents"?

File: chapters/image_accceses.md [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+An `OpTypeImage` can be thought of as a handle to the underlying texel data. To help illustrate this, here is a small program that writes to an image
+
+```glsl
+layout(set = 0, binding = 1, rgba8) uniform writeonly image2D image;
+void main() {
+    imageStore(image, ivec2(0, 0), vec4(1.0));
+}
+```
+
+Taking a closer look at the `OpVariable` we see it is of type [UniformConstant](../chapters/storage_class.md#uniformconstant) which is suppose to be "read-only" (yet there is a clear `OpImageWrite` occuring here).
```
Comment (bashbaug) [inline chapters/image_accceses.md:31]: Nitpick: it's better to say that the variable is in the UniformConstant storage class, or that it points to the UniformConstant storage class.  The actual type of the OpVariable is an OpTypePointer.

https://godbolt.org/z/Md6WE7qEx

File: chapters/image_accceses.md [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+}
+```
+
+The above `OpVariable` is read-only because we are not allowed to adjust its format, width, or multisampling state ever.
+
+# Sampled Images
+
+When using a sampler, you must have an `OpTypeSampledImage` object, this will be directly pointing to an image object of `OpTypeImage`.
+
+To access the sampled image you will use an `OpImage*Sample*` instruction (ex. `OpImageSampleImplicitLod`) to get an `OpSampledImage`. (It can also use an `OpImage*Gather` as well)
```
Comment (bashbaug) [inline chapters/image_accceses.md:56]: This doesn't quite seem right.

You use the `OpSampledImage` instruction to create a sampled image, from an image and a sampler.  This gives you an object of type `OpTypeSampledImage`.

After you create a sampled image, you can access that sampled image with `OpImage*Sample*`, e.g. `OpImageSampleImplicitLod`), but you need to create the sampled image first.

File: chapters/image_accceses.md [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+
+![image_access_sampled_image.png](../images/image_access_sampled_image.png)
+
+## Fetch
+
+There are also `OpImageFetch` (and `OpImageSparseFetch`) instructions which work similar as `OpImage*Sample*`.
+
+The `OpImageFetch` only takes integer texel coordinates. The texture data is read directly without any sampling operations.
+
+This means an `OpImageFetch` will directly access through an `OpImage` and doesn't need an `OpSampledImage` object.
```
Comment (bashbaug) [inline chapters/image_accceses.md:66]: I think this sentence would be clearer if it added a few words to differentiate between instructions and types (or objects of those types).  Something like:

```suggestion
This means the `OpImageFetch` instruction will directly access through an object of type `OpTypeImage` and doesn't need an `OpTypeSampledImage` object.
```

File: chapters/image_accceses.md [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+
+![image_access_storage_image.svg](../images/image_access_storage_image.svg)
+
+## Loading handle vs loading texel
+
+The following code/diagram shows an example how an `OpImageWrite` instruction is used.
+
+![image_access_handle.svg](../images/image_access_handle.svg)
+
+From the code, we se that the `OpImageWrite` calls `OpLoad` to load the the handle (`image_view_a`/`image_view_b`).
```
Comment (bashbaug) [inline chapters/image_accceses.md:83]: typo: 
```suggestion
From the code, we see that the `OpImageWrite` calls `OpLoad` to load the the handle (`image_view_a`/`image_view_b`).
```

File: chapters/image_accceses.md [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+# OpImage
+
+With many other things also starting with `OpImage*` it is easy to lose track what `OpImage` is for. As the spec says simply:
+
+>  [OpImage] Extract the image from a sampled image.
+
+The `OpImage` instruction extracts image object of `OpTypeImage` from an `OpTypeSampledImage`.
+The `OpImage` is describing an action, rather than the input to or the output from an action.
+
+This means you will **not** be using an `OpImage` when accessing an image without a sampler.
```
Comment (bashbaug) [inline chapters/image_accceses.md:100]: This is perhaps slightly oversimplified.  If you are starting from an object of type `OpTypeImage` then you already have what you need to access an image without a sampler.  If you are starting with an object of type `OpTypeSampledImage` though, you need a way to extract the object of type `OpTypeImage` from it, in which case you will need to use `OpImage`.

Maybe:

```suggestion
The `OpImage` instruction is useful if you have an object of type `OpTypeSampledImage` and you need an object of type `OpTypeImage`, say to access the image data without a sampler.
```

**Review by bashbaug — APPROVED** [review]
I'm fine with this going in as-is, but here are some comments and suggestions for future upgrades.  Thanks again for writing this up!

File: chapters/image_accceses.md (hunk 14 above)
Comment (spencer-lunarg) [reply chapters/image_accceses.md:31]: > https://godbolt.org/z/Md6WE7qEx

:100:

File: chapters/image_accceses.md (hunk 12 above)
Comment (spencer-lunarg) [reply chapters/image_accceses.md:12]: added here now

File: chapters/image_accceses.md (hunk 15 above)
Comment (spencer-lunarg) [reply chapters/image_accceses.md:56]: re-worded

**Comment by spencer-lunarg** [comment]
I've applied the various feedback, I think this has had enough time to acquire feedback and going to merge now, like all documentation, it is a living, breathing thing and feel what we have here is better then not having anything and always open for constant improvements to this topic
